hie-plugin-api/Haskell/Ide/Engine/GhcModuleCache.hs

```diff
@@ -99,22 +99,42 @@ data LookupCradleResult = ReuseCradle | LoadCradle CachedCradle | NewCradle File
 -- via 'setCurrentCradle' before the Cradle can be cached via 'cacheCradle'.
 lookupCradle :: FilePath -> GhcModuleCache -> LookupCradleResult
 lookupCradle fp gmc =
-  case currentCradle gmc of
-    Just (dirs, _c) | (any (\d -> d `isPrefixOf` fp) dirs) -> ReuseCradle
-    _ -> case T.match (cradleCache gmc) (B.pack fp) of
-           Just (_k, c, _suf) -> LoadCradle c
-           Nothing  -> NewCradle fp
-
-data CachedCradle = CachedCradle BIOS.Cradle HscEnv
+  lookupInCache fp gmc (const $ const ReuseCradle) LoadCradle $ NewCradle fp
+
+lookupComponentOptions
+  :: HasGhcModuleCache m => FilePath -> m (Maybe BIOS.ComponentOptions)
+lookupComponentOptions fp = do
+  gmc <- getModuleCache
+  return $ lookupInCache fp gmc (const Just) (Just . compOpts) Nothing
+
+lookupInCache
+  :: FilePath
+  -> GhcModuleCache
+  -> (BIOS.Cradle -> BIOS.ComponentOptions -> a)
+  -> (CachedCradle -> a)
+  -> a
+  -> a
+lookupInCache fp gmc cur cached def = case currentCradle gmc of
+  Just (dirs, c, co) | any (`isPrefixOf` fp) dirs -> cur c co
+  _ -> case T.match (cradleCache gmc) (B.pack fp) of
+    Just (_k, c, _suf) -> cached c
+    Nothing            -> def
+
+data CachedCradle = CachedCradle
+  { ccradle :: BIOS.Cradle
+  , hscEnv :: HscEnv
+  , compOpts :: BIOS.ComponentOptions
+  }
 
 instance Show CachedCradle where
-  show (CachedCradle x _) = show x
+  show (CachedCradle x _ _) = show x
 
 data GhcModuleCache = GhcModuleCache
   { cradleCache :: !(T.Trie CachedCradle)
               -- ^ map from FilePath to cradles
+              -- May not include currentCradle
```

Comment:
Why not? The cradle cache should also contain current cradle, no?

Reply:
No idea, I added this as a warning, because I had expected it to be in the cache.